src/optimizer/expr_simplification.ts

```diff
@@ -0,0 +1,713 @@
+                                idText(decl.name),
+                            )!.ast;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "field_decl": {
+                            const newCode = newFieldsMap.get(
+                                idText(decl.name),
+                            )!.ast;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "receiver": {
+                            const newCode = newReceiversMap.get(decl.id)!.ast;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                    }
+                }
+
+                newAst = cloneAstNode({
+                    ...t.ast,
+                    declarations: newDeclarations,
+                });
+
+                break;
+            }
+            case "contract": {
+                const newDeclarations: AstContractDeclaration[] = [];
+
+                for (const decl of t.ast.declarations) {
+                    switch (decl.kind) {
+                        case "asm_function_def":
+                        case "function_def": {
+                            const newCode = newMethods.get(idText(decl.name))!
+                                .ast as AstContractDeclaration;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "constant_def": {
+                            const newCode = newConstantsMap.get(
+                                idText(decl.name),
+                            )!.ast as AstContractDeclaration;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "field_decl": {
+                            const newCode = newFieldsMap.get(
+                                idText(decl.name),
+                            )!.ast;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "receiver": {
+                            const newCode = newReceiversMap.get(decl.id)!.ast;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "contract_init":
+                            newDeclarations.push(newInitializer!.ast);
+                            break;
+                    }
+                }
+
+                newAst = cloneAstNode({
+                    ...t.ast,
+                    declarations: newDeclarations,
+                });
+
+                break;
+            }
+        }
+
+        newTypes.set(t.name, {
+            ...t,
```

Comment:
Spread operator typecheck incorrectly on union types. All the fields that are not modified by the code should be listed explicitly.

The only valid use-case for spread operator is when the type of an object is `Record<string, ...>`. Otherwise the program might compile, but work incorrectly.